### PR TITLE
Backport fixes for ARM64EC import libraries

### PR DESCRIPTION
**Pull request by dpaoliello**
Backports fixes to be able to generate import libraries for ARM64EC:
* <https://github.com/llvm/llvm-project/pull/78769>
* <https://github.com/llvm/llvm-project/pull/81059>
* <https://github.com/llvm/llvm-project/pull/78772>
* <https://github.com/llvm/llvm-project/pull/84834>

There is one modification to these commits: we removed support for `EXPORTAS` in def files to maintain ABI compatibility for the `COFFShortExport` struct.

### CODE REVIEW

**Comment by efriedma-quic** [comment]
This seems like a pretty big change to backport... how useful is it in practice?  I was under the impression that arm64ec lld support is still immature... and if you're using the MSVC linker, you might as well use the MSVC lib/dlltool.

**Comment by dpaoliello** [comment]
Blocking to wait for <https://github.com/llvm/llvm-project/pull/84834>

**Comment by cjacek** [comment]
I have similar feeling to @efriedma-quic . There is probably not much use of those fixes on their own. They are a step towards stand-alone toolchain and are required for the linker, both for its libs part and to generate test cases, but they alone are not enough. They will be more useful when paired with linker support, which I hope will be in a useful shape for LLVM 19. Until then MSVC is required anyway.

**Comment by dpaoliello** [comment]
Sorry, I've got to disagree with this: this is blocking Rust's use of Arm64EC since it can generate import libraries on-the-fly using this functionality (rather than requiring that users install the Windows SDK).

I've tested Arm64EC with Rust, and this is the only blocking issue that I've found.

**Comment by dpaoliello** [comment]
> So your use-case is basically equivalent to using llvm-dlltool, except not using the text parser?

Yep.

> If this is actually enough to make Rust targets usable, then I guess we could consider it, but the fixes aren't structured in a way to make it obvious this won't impact non-ARM64EC targets.

That's fair enough.

@cjacek you have more experience in this area than I do - what's your confidence level with backporting these changes?

**Comment by cjacek** [comment]
It's great to hear that it's enough got Rust. I re-reviewed those commits for impact on non-ARM64EC target and I think it's safe enough.

One thing we could consider is to skip .def file parser part of it. The main downside of it would be that it would require skipping some tests too. Other than than that, it's unlikely to matter and I think that Rust calls `writeImportLibrary` directly anyway.

**Comment by dpaoliello** [comment]
<https://github.com/llvm/llvm-project/pull/84834> has been merged into master and cherry-picked here, so this is ready to land.

> One thing we could consider is to skip .def file parser part of it. 

I'd rather not modify commits since they've already been tested in master and it makes subsequent cherry-picks more difficult.

**Comment by dpaoliello** [comment]
@cjacek if you really want me to revert the def parser portion, then I can do that. Otherwise, can you please sign off?

**Review by cjacek — APPROVED** [review]
I think it's fine to keep it, thanks.

File: llvm/include/llvm/Object/COFFImportFile.h [hunk 1]
```diff
@@ -91,6 +120,10 @@ struct COFFShortExport {
   /// file, this is "baz" in "EXPORTS\nfoo = bar == baz".
   std::string AliasTarget;
 
+  /// Specifies EXPORTAS name. In a .def file, this is "bar" in
+  /// "EXPORTS\nfoo EXPORTAS bar".
+  std::string ExportAs;
```
Comment (nikic) [inline llvm/include/llvm/Object/COFFImportFile.h:125]: This looks like an ABI break.

Comment (dpaoliello) [reply llvm/include/llvm/Object/COFFImportFile.h:125]: If that's problematic, this specific part of the change isn't needed for Rust's scenario.

**Comment by dpaoliello** [comment]
@tstellar I'd like to get this merged into the v18 release branch, but there's some discussion of whether we should trim the changes down to just what is needed or leave them as they were merged into master. Can you please let me know what your preference is?

File: llvm/include/llvm/Object/COFFImportFile.h (hunk 1 above)
Comment (tstellar) [reply llvm/include/llvm/Object/COFFImportFile.h:125]: If you can remove this part of the patch, that would be great.

Comment (dpaoliello) [reply llvm/include/llvm/Object/COFFImportFile.h]: Done

**Pull request merged**